cmd/ceremony/main.go

```diff
@@ -22,6 +23,156 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
+// issueLintCert issues a linting certificate from a given template certificate
+// signed by a given issuer and returns the certificate or an error. The public
+// key from the loaded certificate is checked by the GoodKey package.
+func issueLintCert(tbs, issuer *x509.Certificate, subjectPubKey crypto.PublicKey, signer crypto.Signer, skipLints []string) (*x509.Certificate, error) {
+	lintCertBytes, err := linter.Check(tbs, subjectPubKey, issuer, signer, skipLints)
+	if err != nil {
+		return nil, fmt.Errorf("certificate failed pre-issuance lint: %w", err)
+	}
+
+	lintCert, err := x509.ParseCertificate(lintCertBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	err = kp.GoodKey(context.Background(), lintCert.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	return lintCert, nil
+}
+
+func equalPubKeys(a, b interface{}) bool {
+	aBytes, err := x509.MarshalPKIXPublicKey(a)
+	if err != nil {
+		return false
+	}
+	bBytes, err := x509.MarshalPKIXPublicKey(b)
+	if err != nil {
+		return false
+	}
+	return bytes.Equal(aBytes, bBytes)
+}
+
+func openSigner(cfg PKCS11SigningConfig, pubKey crypto.PublicKey) (crypto.Signer, *hsmRandReader, error) {
```

Comment:
Note that it's still quite difficult to review this diff, because these functions moved from below crlConfig to up here. I've made a local diff that moves these functions back to their original locations to facilitate this review.

Reply:
I will move them back to where they belong.